app/models/waste_exemptions_engine/transient_address.rb

```diff
@@ -53,71 +39,104 @@ def self.create_from_grid_reference_data(data, address_type)
       TransientAddress.create(data)
     end
 
-    private_class_method def self.add_site_details(data, mode)
-      # Add x & y dependent on how the site was entered
-      data = update_xy_from_grid_reference(data) if mode == TransientAddress.modes[:auto]
-      data = update_xy_from_postcode(data) if mode == TransientAddress.modes[:manual]
-
-      # Add the grid reference for sites entered using an address
-      data = update_grid_reference_from_xy(data) unless mode == TransientAddress.modes[:auto]
-
-      # Add the EA administrative area
-      data = update_area_from_xy(data)
-
-      data
+    private
+
+    # Disabling the following cops for these reasons.
+    #
+    # The part of the code that tips the balance to being too complex is the
+    # `if foo.blank? checks after each method call. Take them away and we're
+    # fine.
+    #
+    # The first issue is if we move those checks into the methods, then the
+    # complexity warnings move to them instead. We would argue that actually
+    # neither this method, nor the methods we call here are actually that
+    # complex and it would make things worse not better to break things up
+    # further.
+    #
+    # Secondly we have a nice separation between what the checks represent. Here
+    # our checks are determining whther we need to run the updates or not. The
+    # update methods themselves don't care if the fields they are updating are
+    # populated or not, they just care whether they have the data they need to
+    # run.
+    #
+    # By splitting them up in this way we retain the flexibility to call these
+    # methods to update the fields irrespectibe of their current state. And
+    # we reduce what they need to know to just 'have i got the data I need'.
+    #
+    # rubocop:disable Metrics/CyclomaticComplexity
+    # rubocop:disable Metrics/PerceivedComplexity
+    def update_site_details
+      return unless site?
+
+      update_x_and_y_from_grid_reference if x.blank? || y.blank?
+      update_x_and_y_from_postcode if x.blank? || y.blank?
+      update_grid_reference_from_x_and_y if grid_reference.blank?
+      update_area_from_x_and_y if area.blank?
+
+      save!
```

Comment:
Can we create a service that deal with this stuff? Is a lot of code for this model to deal with.

Reply:
😩 Yes boss!

![poodr](https://user-images.githubusercontent.com/1789650/66472901-20e66c80-ea86-11e9-9cf9-a4c3b205d9b3.gif)